docs/blog/2019-02-04-pragmatic-lessons-from-converting-to-react-hooks/index.md

```diff
@@ -0,0 +1,127 @@
+  }
+
+  return {
+    currIndex,
+    isPlaying,
+    setIsPlaying,
+    updateSlide,
+  }
+}
+```
+
+The benefits of this structural change go beyond aesthetic cleanup. One major benefit is the cleanup and re-running of useEffect when specific properties change. This eliminates the need of storing the timeout id as a static property. It also eliminates imperative prop comparisons in `componentDidUpdate`. Instead, the function declares the data that will trigger a re-run of this side effect. If `currIndex` or `isPlaying` change, the effect will re-run. First the function will run the cleanup function to clear the timeout, and then will run the effect again. This compounds when a component requires many side effects.
+
+The `useEffect` hook brings more clarity and organization to the side effects in your components. Class based components force you to provide your set up and tear down across three separate life cycle methods: `componentDidMount`, `componentDidUpdate`, and `componentWillUnmount`. This often results in duplication of code. A component may use many side effects which are all placed in these same lifecycle methods. The `useEffect` Hook shifts this paradigm by isolating the functionality of a given side effect in one function. Clean up for the effect is set as a return function. This colocation allows for easier mental parsing and grouping functions by their area of concern.
+
+An extra benefit of `useState` is the ability to reference the current state without additional syntax. With Hooks you can call `setCurrIndex((currIndex + 1) % slides.length)` rather than passing a function which returns the partial state object to update. This results in cleaner code, and eliminated the need for the side-loaded function for the Slideshow component. Converting the Slideshow from a render prop component to a custom Hook resulted in a net reduction of 76 lines of code. Other components within the site utilized class components to manage a single piece of state. Hooks allow you to reduce the boilerplate of a class based component in these circumstances. Functional components connect to state with one line of code.
+
+## Structuring Hooks in Gatsby
+
+Currently I am structuring all my custom hooks in a top level folder so that other components can import them from a central location. Pulling the React dependencies out of Gatsby in version 2 allows for using Hooks immediately. To begin using Hooks today, update React and React-DOM to 16.8.0. There are considerations to take with the `useEffect Hook`. If it references the window object, you need to check that window is defined to avoid Gatsby build errors. These effects would normally live in `componentDidMount` where the component hydrates in the DOM. Hooks are called in the build process. The Gatsby docs have great resources for [debugging HTML builds](docs/debugging-html-builds/) if you encounter this issue.
+
+```javascript
+function useMediaQuery() {
```

Comment:
(Note: I swapped this a bit. I like to use min-width queries rather than max-width queries!)

Reply:
Definitely. This example was more simple since it fit the use case for this particular project.

One thing I have thought of doing on another project which has more complex needs is to have the useMediaQuery function accept an object where some common queries would check if the corresponding options key was set to true. If so it would add the corresponding listener. If false, it would not add the listener. There could potentially be optional keys for adding custom queries as well.